sled-agent/config-reconciler/src/reconciler_task/zones.rs

```diff
@@ -1035,6 +1264,149 @@ fn config_differs_only_by_image_source(
+        image_source: &OmicronZoneImageSource,
+        internal_disks: &InternalDisks,
+    ) -> OmicronZoneFileSource {
+        match image_source {
+            OmicronZoneImageSource::InstallDataset => {
+                let file_name = zone_kind.artifact_in_install_dataset();
+
+                // There's always at least one image path (the RAM disk below).
+                let mut search_paths = Vec::with_capacity(1);
+
+                // Inject an image path if requested by a test.
+                if let Some(path) = &self.image_directory_override {
+                    search_paths.push(path.clone());
+                };
+
+                // Any zones not part of the RAM disk are managed via the zone
+                // manifest.
+                let hash = install_dataset_hash(
+                    log,
+                    self,
+                    zone_kind,
+                    internal_disks,
+                    |path| search_paths.push(path),
+                );
+
+                // Look for the image in the RAM disk as a fallback. Note that
+                // install dataset images are not stored on the RAM disk in
+                // production, just in development or test workflows.
+                search_paths.push(Utf8PathBuf::from(RAMDISK_IMAGE_PATH));
+
+                OmicronZoneFileSource {
+                    location: OmicronZoneImageLocation::InstallDataset { hash },
+                    file_source: ZoneImageFileSource {
+                        file_name: file_name.to_owned(),
+                        search_paths,
+                    },
+                }
+            }
+            OmicronZoneImageSource::Artifact { hash } => {
+                // TODO: implement mupdate override here.
+                //
+                // Search both artifact datasets. This iterator starts with the
+                // dataset for the boot disk (if it exists), and then is followed
+                // by all other disks.
+                let search_paths =
+                    internal_disks.all_artifact_datasets().collect();
+                OmicronZoneFileSource {
+                    // TODO: with mupdate overrides, return InstallDataset here
+                    location: OmicronZoneImageLocation::Artifact {
+                        hash: Ok(*hash),
+                    },
+                    file_source: ZoneImageFileSource {
+                        file_name: hash.to_string(),
+                        search_paths,
+                    },
+                }
+            }
+        }
+    }
+}
+
+fn install_dataset_hash<F>(
+    log: &slog::Logger,
+    resolver_status: &ResolverStatus,
+    zone_kind: ZoneKind,
+    internal_disks: &InternalDisks,
+    mut search_paths_cb: F,
+) -> Result<ArtifactHash, ZoneImageLocationError>
+where
+    F: FnMut(Utf8PathBuf),
+{
+    // XXX: we ask for the boot zpool to be passed in here. But
+    // `AllZoneImages` also caches the boot zpool. How should we
```

Comment:
Is this comment still relevant? `AllZoneImages` isn't otherwise referenced in this file anymore, I think?